### PR TITLE
FIXME in nodeSort.c: re-implement explain analyze related code in tuplesort

### PR DESCRIPTION
**Pull request by yaowangm**
### Brief

The FIXME is about re-implementing explain analyze related codes in tuplesort. In tuplesort.c on GPDB6, there are some codes to collect statistic info for "explain analyze" output. On GPDB7, most of codes were rewritten and the explain analyze related codes were gone. The task is to re-implement them.

### Code

The major logic is straightforward and follows similar logic on GPDB6 to collect info and calculate statistic numbers in tuplesort. An exception is, since the code path is pretty different between GPDB6 and 7, I have to add a function `tuplesort_finalize_stats()` (different from GPDB6 version) to handle the stuffs when statistic info can be finalized.

### Test

The code change impacts a lot of test cases. In addition, in some (not all) cases we compare some extracted info from PLAN as "explain analyze" output, but not the PLAN itself. It means, the extra info added by this code change may be ignored by the cases. To check the extra info, we need:

1. Use UDF to detect particular info in PLAN: I added two UDFs to 
src/test/regress/sql/setup.sql: `get_explain_output()` to run SQL query with "explain ..." and return result as text rows, and `extract_plan_stats()` to detect particular keywords and return statistic info as a table. Both are placed on schema `test_util` for further reuse in future. 
2. Share common codes among mutilple test cases: as said, common codes were added into setup.sql. Similar to isolation2, I added "setup" to `setup_tests` in pg_regress_main.c, so `setup.sql` will be called automatically before running any test case. In addition, `pg_regress` might be called outside of the source tree by other test cases, and there might not be test case "setup" on the dir. I updated code `in run_single_test()` to detect whether there is `setup.sql`, and ignore the calling if not.

Please note:

1. Some tests implemented similar UDFs to detect particular info but not for common use. Theoretically they can be replaced by the common version. I prefer to keep them as-is and use the common version in future code. 
2. Some tests adapts `setup.sql` as prerequisite by hard-coding in script (e.g. `contrib/extprotocol`). Since we have an automatically running version, I have to update the cases by removing the hard-coding or something.

### CODE REVIEW

**Comment by yaowangm** [comment]
### Brief

The FIXME is about re-implementing explain analyze related codes in tuplesort. In tuplesort.c on GPDB6, there are some codes to collect statistic info for "explain analyze" output. On GPDB7, most of codes were rewritten and the explain analyze related codes were gone. The task is to re-implement them.

### Code

The major logic is straightforward and follows similar logic on GPDB6 to collect info and calculate statistic numbers in tuplesort. An exception is, since the code path is pretty different between GPDB6 and 7, I have to add a function `tuplesort_finalize_stats()` (different from GPDB6 version) to handle the stuffs when statistic info can be finalized.

### Test

The code change impacts a lot of test cases. In addition, in some (not all) cases we compare some extracted info from PLAN as "explain analyze" output, but not the PLAN itself. It means, the extra info added by this code change may be ignored by the cases. To check the extra info, we need:

1. Use UDF to detect particular info in PLAN: I added two UDFs to 
src/test/regress/sql/setup.sql: `get_explain_output()` to run SQL query with "explain ..." and return result as text rows, and `extract_plan_stats()` to detect particular keywords and return statistic info as a table. Both are placed on schema `test_util` for further reuse in future. 
2. Share common codes among mutilple test cases: as said, common codes were added into setup.sql. Similar to isolation2, I added "setup" to `setup_tests` in pg_regress_main.c, so `setup.sql` will be called automatically before running any test case. In addition, `pg_regress` might be called outside of the source tree by other test cases, and there might not be test case "setup" on the dir. I updated code `in run_single_test()` to detect whether there is `setup.sql`, and ignore the calling if not.

Please note:

1. Some tests implemented similar UDFs to detect particular info but not for common use. Theoretically they can be replaced by the common version. I prefer to keep them as-is and use the common version in future code. 
2. Some tests adapts `setup.sql` as prerequisite by hard-coding in script (e.g. `contrib/extprotocol`). Since we have an automatically running version, I have to update the cases by removing the hard-coding or something.

File: src/backend/utils/sort/tuplesort.c [hunk 1]
```diff
@@ -2971,6 +3001,7 @@ dumptuples(Tuplesortstate *state, bool alltuples)
 {
 	int			memtupwrite;
 	int			i;
+	long		spilledBytes = state->availMem;
```
Comment (adam8157) [inline src/backend/utils/sort/tuplesort.c:3004]: Is there a reason to assign `spilledBytes`?

File: src/backend/utils/sort/tuplesort.c [hunk 2]
```diff
@@ -4674,3 +4716,41 @@ free_sort_tuple(Tuplesortstate *state, SortTuple *stup)
...
+ * Finalize the EXPLAIN ANALYZE stats.
+ */
+void
+tuplesort_finalize_stats(Tuplesortstate *state,
+					TuplesortInstrumentation *stats)
+{
+	if(state->instrument)
+	{
+		state->instrument->workmemused = MemoryContextGetPeakSpace(state->sortcontext);
+		state->instrument->execmemused += MemoryContextGetPeakSpace(state->sortcontext);
```
Comment (adam8157) [inline src/backend/utils/sort/tuplesort.c:4753]: Why that one is `=` but the other is `+=`?

File: contrib/extprotocol/Makefile [hunk 3]
```diff
@@ -1,6 +1,6 @@
 MODULE_big = gpextprotocol
 OBJS       = gpextprotocol.o
-REGRESS = setup exttableext
+REGRESS = exttableext
```
Comment (adam8157) [inline contrib/extprotocol/Makefile:3]: For other reviewers' info, this is fine, extprotocol is a gp-only thing.

File: src/test/gpdb_pitr/test_gpdb_pitr.sh [hunk 4]
```diff
@@ -155,9 +155,15 @@ echo "Restarting cluster now that the new cluster is properly configured..."
 export COORDINATOR_DATA_DIRECTORY=$REPLICA_MASTER
 gpstop -ar
 
+# Temporarily rename setup.sql to avoid that it will be invoked by pg_regress automatically
+mv ./sql/setup.sql ./sql/setup.bak
```
Comment (adam8157) [inline src/test/gpdb_pitr/test_gpdb_pitr.sh:159]: Why is this needed?

**Comment by kainwen** [comment]
Not look into the code yet.

Just comments on the test: from my side, can we just add a single file to test the piece of code, and leave most of other case untouched? This PR changes too many files, I doubt it is necassary.

**Comment by yaowangm** [comment]
> Not look into the code yet.
> 
> Just comments on the test: from my side, can we just add a single file to test the piece of code, and leave most of other case untouched? This PR changes too many files, I doubt it is necassary.

The code change generates more output of Sort operator in PLAN, so all test cases including Sort in PLAN were impacted. We don't need to add more test case since all the scenarios to be tested have been existing already.

There are three kinds of test cases:

1. For some test cases which compare PLAN text directly, we need to update output files anyway.
2. For some test cases which compare extracted info of PLAN, theoretically we can just add UDFs and relevant results to some of them and ignore others without breaking the cases. However, it means we might ignore some scenarios which don't generate expected info and they might be difficult to be detected.
3. Some test cases were impacted because of "automatically running setup" and need to be updated anyway.

So I prefer to update all relevant test cases.

**Comment by kainwen** [comment]
> For some test cases which compare PLAN text directly, we need to update output files anyway.

Can you give an example? Seems I do not see any before.

> However, it means we might ignore some scenarios which don't generate expected info and they might be difficult to be detected.

Upstream does not have such kind of test. We just need test to test the added code.

> Some test cases were impacted because of "automatically running setup" and need to be updated anyway.

We can get rid of "automatically running setup", and only in the newly added case create those UDFs.

**Comment by haolinw** [comment]
Recent upstream has test_setup.sql seems to do similar work, may be as a reference for this.
Prefer to keeping consistent with upstream as much as possible in src/test/regress for future merging code smoothly.

**Comment by yaowangm** [comment]
> > For some test cases which compare PLAN text directly, we need to update output files anyway.
> 
> Can you give an example? Seems I do not see any before.
> 
Please see dpe.sql.

> > However, it means we might ignore some scenarios which don't generate expected info and they might be difficult to be detected.
> 
> Upstream does not have such kind of test. We just need test to test the added code.
> 
The code change impacts all Sort related test cases, including not only GP specific but also upstream. But it's not the point. The point is we need to guarantee that the code works for all known scenarios. I did hit a test case which didn't generate expected info because of a code path I was not aware of. That's the values of test cases, right?

> > Some test cases were impacted because of "automatically running setup" and need to be updated anyway.
> 
> We can get rid of "automatically running setup", and only in the newly added case create those UDFs.

If that, we have to add setup to all schedules and other scripts calling regress tests. That should be very ugly and error-prone.

**Comment by yaowangm** [comment]
> Recent upstream has test_setup.sql seems to do similar work, may be as a reference for this. Prefer to keeping consistent with upstream as much as possible in src/test/regress for future merging code smoothly.

That should be great if it's an option. I will take a look. Thanks.

**Comment by yaowangm** [comment]
I studied test_setup in recent PG: https://github.com/postgres/postgres/blob/master/src/test/regress/sql/test_setup.sql

It's not so ideal because it's implemented by sort of hardcoding in schedule files. To apply the approach on GP, we need to add test_setup to proper places of all schedule files and maintain the dependencies, which is what I wanted to avoid by "automatically running setup". However, I agree to keep consistent with upstream as much as possible. It's best if we can push the "automatically running setup" to upstream, but I think it's another topic.

I have updated code accordingly and pipeline is clean now. Please take another look.

File: src/backend/utils/sort/tuplesort.c (hunk 1 above)
Comment (adam8157) [inline src/backend/utils/sort/tuplesort.c:3004]: The logic is correct, but the name is confusing.

How about `long prevAvailMem = state->availMem;`

File: src/backend/utils/sort/tuplesort.c [hunk 5]
```diff
@@ -3080,6 +3111,13 @@ dumptuples(Tuplesortstate *state, bool alltuples)
 			 pg_rusage_show(&state->ru_start));
 #endif
 
+	/* CDB: Accumulate total size of spilled tuples. */
+	spilledBytes = state->availMem - spilledBytes;
```
Comment (adam8157) [inline src/backend/utils/sort/tuplesort.c:3115]:
```
if (state->availMem > prevAvailMem)
    state->spilledBytes += (state->availMem - prevAvailMem);
```

**Review by adam8157 — COMMENTED** [review]
Others look good to me.

Please squash your commits, update the commit message (setup.sql behavior changed), and rebase.

nitpick, `Re-implement explain analyze related code in tuplesort` is good enough as the subject.
